src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,488 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark.lightgbm.dataset
+
+import com.microsoft.ml.spark.lightgbm.ColumnParams
+import com.microsoft.ml.spark.lightgbm.swig.{BaseSwigArray, ChunkedArray, DoubleChunkedArray, DoubleSwigArray,
+  FloatChunkedArray, FloatSwigArray, IntChunkedArray, IntSwigArray}
+import org.apache.spark.sql.Row
+
+import scala.collection.mutable.ListBuffer
+
+object ChunkedArrayUtils {
+  def copyChunkedArray[T](chunkedArray: ChunkedArray[T],
+                          mainArray: BaseSwigArray[T],
+                          threadRowStartIndex: Long,
+                          chunkSize: Long): Unit = {
+    // Copy in parallel on each thread
+    // First copy full chunks
+    val chunkCount = chunkedArray.getChunksCount() - 1
+    for (chunk <- 0L until chunkCount) {
+      for (in_chunk_idx <- 0L until chunkSize) {
+        mainArray.setItem(threadRowStartIndex + chunk * chunkSize + in_chunk_idx,
+          chunkedArray.getItem(chunk, in_chunk_idx, (-1).asInstanceOf[T]))
+      }
+    }
+    // Next copy filled values from last chunk only
+    val lastChunkCount = chunkedArray.getLastChunkAddCount()
+    for (last_chunk_idx <- 0L until lastChunkCount) {
+      mainArray.setItem(threadRowStartIndex + chunkCount * chunkSize + last_chunk_idx,
+        chunkedArray.getItem(chunkCount, last_chunk_idx, (-1).asInstanceOf[T]))
+    }
+  }
+}
+
+trait AggregatedColumns
+
+class BaseChunkedColumns(columnParams: ColumnParams,
+                         chunkSize: Int) extends AggregatedColumns {
+  val labelsChunkedArray = new FloatChunkedArray(chunkSize)
+  val weightChunkedArrayOpt = columnParams.weightColumn.map {
+    _ => new FloatChunkedArray(chunkSize)
+  }
+  val initScoreChunkedArrayOpt = columnParams.initScoreColumn.map {
+    _ => new DoubleChunkedArray(chunkSize)
+  }
+  val groupColumnValues: ListBuffer[Row] = new ListBuffer[Row]()
+
+  var rowCount = 0
+
+  def release(): Unit = {
+    // Clear memory
+    labelsChunkedArray.delete()
+    weightChunkedArrayOpt.foreach(_.delete())
+    initScoreChunkedArrayOpt.foreach(_.delete())
+  }
+}
+
+class SparseChunkedColumns(columnParams: ColumnParams,
```

Comment:
done! I made the change in PR https://github.com/Azure/mmlspark/pull/1066 since this PR will be abandoned.
I made all of the new classes private and many of them final.